influxql/parser.go

```diff
@@ -265,6 +267,59 @@ func (p *Parser) parseSetPasswordUserStatement() (*SetPasswordUserStatement, err
 	return stmt, nil
 }
 
+// parseBackfillStatement parses the backfill statement. It assumes that the
+// BACKFILL keyword has already been consumed.
+func (p *Parser) parseBackfillStatement() (*BackfillStatement, error) {
+	stmt := &BackfillStatement{}
+	// Parse the CQ name
+	ident, err := p.parseIdent()
+	if err != nil {
+		return nil, err
+	}
+	stmt.QueryName = ident
+
+	// get the ON keywork
+	tok, pos, lit := p.scanIgnoreWhitespace()
+	if tok != ON {
+		return nil, newParseError(tokstr(tok, lit), []string{"ON"}, pos)
+	}
+
+	// Get the database name
+	ident, err = p.parseIdent()
+	if err != nil {
+		return nil, err
+	}
+	stmt.Database = ident
+
+	// get the FROM keyword
+	tok, pos, lit = p.scanIgnoreWhitespace()
+	if tok != FROM {
+		return nil, newParseError(tokstr(tok, lit), []string{"FROM"}, pos)
+	}
+
+	expr, err := p.ParseExpr()
+	if err != nil {
+		return nil, err
+	}
+	stmt.From = expr
+
+	// Get the UNTIL keyword. This is optional, so if we see EOF, just ignore it.
+	tok, pos, lit = p.scanIgnoreWhitespace()
+	if tok == EOF {
+		return stmt, nil
+	} else if tok != UNTIL {
+		return nil, newParseError(tokstr(tok, lit), []string{"UNTIL"}, pos)
+	}
```

Comment:
If it's not an `UNTIL` token, it should unscan it and return.  The main loop should handle whatever comes next (could be `EOF` or `SEMICOLON`).